pkg/querier/frontend/downstream_roundtripper.go

```diff
@@ -0,0 +1,40 @@
+package frontend
+
+import (
+	"net/http"
+	"net/url"
+	"path"
+
+	"github.com/opentracing/opentracing-go"
+)
+
+// RoundTripper that forwards requests to downstream URL.
+type downstreamRoundTripper struct {
+	downstreamURL *url.URL
+}
+
+func NewDownstreamRoundTripper(downstreamURL string) (http.RoundTripper, error) {
+	u, err := url.Parse(downstreamURL)
+	if err != nil {
+		return nil, err
+	}
+
+	return &downstreamRoundTripper{downstreamURL: u}, nil
+}
+
+func (d downstreamRoundTripper) RoundTrip(r *http.Request) (*http.Response, error) {
+	tracer, span := opentracing.GlobalTracer(), opentracing.SpanFromContext(r.Context())
+	if tracer != nil && span != nil {
+		carrier := opentracing.HTTPHeadersCarrier(r.Header)
+		err := tracer.Inject(span.Context(), opentracing.HTTPHeaders, carrier)
+		if err != nil {
+			return nil, err
```

Comment:
This can fail for these reasons:
- invalid span context is passed in first argument
- unknown format (`opentracing.HTTPHeaders`) passed as second argument
- invalid combination of format (`opentracing.HTTPHeaders`) and `carrier`

All of these are errors in the code, and cannot happen based on HTTP request itself. If code is correct, this call will never fail (it just sets some HTTP headers). On the other hand, ignoring the errors can mask possible bugs.